File: cc/leetcode/array/907_sum_of_subarray_min.cc
class Solution {
public:
    int sumSubarrayMins(vector<int>& A) {
        long MOD = 1e9 + 7;
        stack<pair<int, int>> s;
        long curr = 0, res = 0;
        for (int i = 0; i < A.size(); ++i) {
            int tmp_cnt = 1;
            long new_ = curr + A[i];
            while (s.size() > 0 && s.top().first >= A[i]) {
                auto old_item = s.top();
                s.pop();
                new_ -= (old_item.first - A[i]) * old_item.second;
                tmp_cnt += old_item.second;
            }
            s.push(make_pair(A[i], tmp_cnt));
            res = (res + new_) % MOD;
            curr = new_ % MOD;
        }
        return res;
    }
};

File: python/leetcode/array/stack/907_sum_of_subarray_min.py
"""
907. Sum of Subarray Minimums (Medium)

Given an array of integers A, find the sum of min(B), where B ranges over 
every (contiguous) subarray of A.

Since the answer may be large, return the answer modulo 10^9 + 7.

Example 1:

Input: [3,1,2,4]
Output: 17
Explanation: Subarrays are [3], [1], [2], [4], [3,1], [1,2], [2,4], 
[3,1,2], [1,2,4], [3,1,2,4]. 
Minimums are 3, 1, 2, 4, 1, 1, 2, 1, 1, 1.  Sum is 17.

Note:

1 <= A.length <= 30000
1 <= A[i] <= 30000
"""

class Solution(object):
    def sumSubarrayMins(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        MOD = 10 ** 9 + 7
        stack = []
        curr = 0
        res = 0
        for i, item in enumerate(A):
            tmp_cnt = 1
            new = curr + item
            while stack and stack[-1][0] >= item:
                tmp, old_cnt = stack.pop()
                new -= (tmp - item) * old_cnt
                tmp_cnt += old_cnt
            stack.append((item, tmp_cnt))
            res = (res + new) % MOD
            curr = new % MOD
        return res


if __name__ == "__main__":
    a = Solution()
    print(a.sumSubarrayMins([3,1,2,4]))
    print(a.sumSubarrayMins([81, 55, 2]))
